### PR TITLE
fix(parsers.csv): Allow null-delimiters

### LINKED ISSUE #12243: Null delimiter cannot be used
### Relevant telegraf.conf

```toml
[[inputs.mqtt_consumer]]
## Topics that will be subscribed to.
  topics = [
    "+/+/measurements"
  ]
  data_format = "csv"
  csv_header_row_count = 0
  csv_column_names = ["Timestamp","Temp1","Temp2","Temp3","Temp4","Temp5","Temp6","Temp7","Temp8","Temp9","Temp10","Temp11","Temp12","Temp13","Freq1","Vibr1","Vibr2"]
  csv_column_types = ["string","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float"]
  csv_delimiter = "\u0000"
  csv_comment = "#"
```

### Logs from Telegraf

```text
telegraf   | 2022-11-15T09:44:09Z I! Using config file: /etc/telegraf/telegraf.conf
telegraf   | 2022-11-15T09:44:09Z I! Starting Telegraf 1.24.2
telegraf   | 2022-11-15T09:44:09Z I! Available plugins: 222 inputs, 9 aggregators, 26 processors, 20 parsers, 57 outputs
telegraf   | 2022-11-15T09:44:09Z I! Loaded inputs: mqtt_consumer
telegraf   | 2022-11-15T09:44:09Z I! Loaded aggregators:
telegraf   | 2022-11-15T09:44:09Z I! Loaded processors:
telegraf   | 2022-11-15T09:44:09Z I! Loaded outputs: influxdb_v2
telegraf   | 2022-11-15T09:44:09Z I! Tags enabled: host=localhost
telegraf   | 2022-11-15T09:44:09Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"localhost", Flush Interval:10s
telegraf   | 2022-11-15T09:44:09Z D! [agent] Initializing plugins
telegraf   | 2022-11-15T09:44:09Z D! [agent] Connecting outputs
telegraf   | 2022-11-15T09:44:09Z D! [agent] Attempting connection to [outputs.influxdb_v2]
telegraf   | 2022-11-15T09:44:09Z D! [agent] Successfully connected to outputs.influxdb_v2
telegraf   | 2022-11-15T09:44:09Z D! [agent] Starting service inputs
telegraf   | 2022-11-15T09:44:09Z I! [inputs.mqtt_consumer] Connected [tcp://mosquitto:1883]
telegraf   | 2022-11-15T09:44:19Z D! [outputs.influxdb_v2] Buffer fullness: 0 / 10000 metrics
telegraf   | 2022-11-15T09:44:27Z E! [inputs.mqtt_consumer] Error in plugin: csv: invalid field or comment delimiter
telegraf   | 2022-11-15T09:44:28Z E! [inputs.mqtt_consumer] Error in plugin: csv: invalid field or comment delimiter
telegraf   | 2022-11-15T09:44:29Z D! [outputs.influxdb_v2] Buffer fullness: 0 / 10000 metrics
telegraf   | 2022-11-15T09:44:29Z E! [inputs.mqtt_consumer] Error in plugin: csv: invalid field or comment delimiter
telegraf   | 2022-11-15T09:44:30Z E! [inputs.mqtt_consumer] Error in plugin: csv: invalid field or comment delimiter
telegraf   | 2022-11-15T09:44:31Z E! [inputs.mqtt_consumer] Error in plugin: csv: invalid field or comment delimiter
```

### System info

Telegraf 1.24.2 (telegraf:lastest image)

### Docker

Docker 20.10.20

### Steps to reproduce

1. Set up config to null delimiters
2. Start up env
3. send csv data separated with  "\u000"
...

### Expected behavior

Like its works with tab ("\u0009"), data parsed and send to influx

### Actual behavior

Throw error : "Error in plugin: csv: invalid field or comment delimiter" 

### Additional info

The device setting cannot be changed , it's sending line data with "\u0000"., one line per cycle, without header.

Thank for your help.

### PR DESCRIPTION
**Pull request by Meceron**
[https://github.com/influxdata/telegraf/issues/12243](url)

- [ ] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [ ] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #12243

Fixing null delimiters in csv, change the selected bytes before parsing to pipeline.

### CODE REVIEW

**Comment by powersj** [comment]
Hi,

Thanks for putting up this PR. It appears that this is essentially a "find and replace" before doing any parsing. Instead of having this hard-code the replacement, can we instead rename `csv_force_replace_delimiter` to `csv_delimiter_replace` and use the `csv_delimiter` as the replacement?

Essentially if I had:

```
csv_delimiter = ","
csv_delimiter_replace = "\u0000"
```

It would replace those null bytes with commas.

What do you think about that? Before you go off and make changes I would like to hear from @srebhan as well

**Comment by srebhan** [comment]
@Meceron thanks for your PR! Instead of forcing the user to give another option how about the following (in pseudo-code):

1. Check if delimiter is valid, i.e. copy the [validDelim function of encoding/csv](https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/encoding/csv/reader.go;l=95)
2. If invalid:
2.1 replace all comma (`, ) with `\ufffd` which is the unicode replacement character and should not appear in the text
2.2 replace all user-specified delimiter instances by comma
2.3 change CSV delimiter to comma
2.4 set a flag that we replaced data e.g. `p.delimiterReplaced = true`
3. do the csv parsing
4. If `p.delimiterReplaced == true` replace `\ufffd`  in _all_ resulting fields back to comma

What do you think?

**Comment by Meceron** [comment]
Thanks for the advice! I've changed code.

File: plugins/parsers/csv/README.md [hunk 1]
```diff
@@ -59,6 +59,8 @@ values.
 
   ## The separator between csv fields
   ## By default, the parser assumes a comma (",")
+  ## Please note that if you use blacklisted delimiters (e.g. "\u0000"), this will be changed to comma.
+  ## in the data the blacklisted delimiters will be changed to comma, and original commas will be changed to "\ufffd"
```
Comment (powersj) [inline plugins/parsers/csv/README.md:63]: What about:

```suggestion
  ## Please note that if you use invalid delimiters (e.g. "\u0000"), commas
  ## will be changed to "\ufffd", the invalid delimiters changed to a comma
  ## during parsing, and afterwards the invalid characters and commas are
  ## returned to their original values.
```

File: plugins/parsers/csv/parser.go [hunk 2]
```diff
@@ -181,22 +188,36 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
...
 		p.Reset()
 	}
-
+	// Replace blacklisted delimiter bytes
+	if p.delimiterReplaced {
+		// replace value commes to replacement bytes
+		buf = bytes.Replace(buf, []byte(commaByte), []byte(replacementByte), -1)
+		// replace delimiters to comma
+		buf = bytes.Replace(buf, []byte(p.Delimiter), []byte(commaByte), -1)
+	}
```
Comment (powersj) [inline plugins/parsers/csv/parser.go:220]:
```suggestion
	// If using an invalid delimiter, replace commas with replacement and
	// invalid delimiter with commas
	if p.invalidDelimiter {
		buf = bytes.Replace(buf, []byte(commaByte), []byte(replacementByte), -1)
		buf = bytes.Replace(buf, []byte(p.Delimiter), []byte(commaByte), -1)
	}
```

File: plugins/parsers/csv/parser.go [hunk 3]
```diff
@@ -181,22 +188,36 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
 	csvReader := csv.NewReader(r)
 	// ensures that the reader reads records of different lengths without an error
 	csvReader.FieldsPerRecord = -1
-	if p.Delimiter != "" {
+	if !p.delimiterReplaced && p.Delimiter != "" {
 		csvReader.Comma = []rune(p.Delimiter)[0]
 	}
+	// if the delimiter is balckited by the go reader
```
Comment (powersj) [inline plugins/parsers/csv/parser.go:194]:
```suggestion
	// Check if delimiter is invalid
```

File: plugins/parsers/csv/parser.go [hunk 4]
```diff
@@ -50,6 +54,8 @@ type Parser struct {
 
 	gotColumnNames bool
 
+	delimiterReplaced bool
```
Comment (powersj) [inline plugins/parsers/csv/parser.go:57]:
```suggestion
	invalidDelimiter bool
```

File: plugins/parsers/csv/parser.go [hunk 5]
```diff
@@ -140,6 +146,7 @@ func (p *Parser) Init() error {
 		if len(runeStr) > 1 {
 			return fmt.Errorf("csv_delimiter must be a single character, got: %s", p.Delimiter)
 		}
+		p.delimiterReplaced = !validDelim(runeStr[0])
```
Comment (powersj) [inline plugins/parsers/csv/parser.go:149]:
```suggestion
		p.invalidDelimiter = !validDelim(runeStr[0])
```

File: plugins/parsers/csv/parser.go [hunk 6]
```diff
@@ -181,22 +188,36 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
 	csvReader := csv.NewReader(r)
 	// ensures that the reader reads records of different lengths without an error
 	csvReader.FieldsPerRecord = -1
-	if p.Delimiter != "" {
+	if !p.delimiterReplaced && p.Delimiter != "" {
```
Comment (powersj) [inline plugins/parsers/csv/parser.go:191]:
```suggestion
	if !p.invalidDelimiter && p.Delimiter != "" {
```

File: plugins/parsers/csv/parser.go [hunk 7]
```diff
@@ -181,22 +188,36 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
...
+}
+
 func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
 	// Reset the parser according to the specified mode
 	if p.ResetMode == "always" {
 		p.Reset()
 	}
-
+	// Replace blacklisted delimiter bytes
+	if p.delimiterReplaced {
```
Comment (powersj) [inline plugins/parsers/csv/parser.go:215]:
```suggestion
	if p.invalidDelimiter {
```

**Review by powersj — COMMENTED** [review]
Thanks for the quick turn around! I have put up some comments that clarify the docs and rename the bool to `invalidDelimiter`.

Take a look and let us know!

**Comment by srebhan** [comment]
@Meceron any news on this PR?

**Review by Hipska — CHANGES_REQUESTED** [review]
Whoops, I didn't finished my review..

File: plugins/parsers/csv/parser.go [hunk 8]
```diff
@@ -23,6 +24,9 @@ import (
 
 type TimeFunc func() time.Time
 
+var replacementByte = "\ufffd"
+var commaByte = "\u002C"
```
Comment (srebhan) [inline plugins/parsers/csv/parser.go:28]: Can you please make this
```suggestion
const replacementByte = "\ufffd"
const commaByte = "\u002C"
```

File: plugins/parsers/csv/parser.go [hunk 9]
```diff
@@ -181,22 +188,35 @@ func (p *Parser) compile(r io.Reader) *csv.Reader {
...
 		csvReader.Comment = []rune(p.Comment)[0]
 	}
 	csvReader.TrimLeadingSpace = p.TrimSpace
 	return csvReader
 }
 
+func validDelim(r rune) bool {
+	return r != 0 && r != '"' && r != '\r' && r != '\n' && utf8.ValidRune(r) && r != utf8.RuneError
+}
+
```
Comment (srebhan) [inline plugins/parsers/csv/parser.go:208]: Please attribute the code here
```suggestion
// Taken from upstream Golang code see
// https://github.com/golang/go/blob/release-branch.go1.19/src/encoding/csv/reader.go#L95
func validDelim(r rune) bool {
	return r != 0 && r != '"' && r != '\r' && r != '\n' && utf8.ValidRune(r) && r != utf8.RuneError
}

```

**Comment by Meceron** [comment]
Sorry , I was on sickleave :(. As I see You All done everything, thank you!

**Review by srebhan — COMMENTED** [review]
I only do have two small comments @Meceron. Can you please address those and consider the other comment of @Hipska!?

**Comment by powersj** [comment]
I wanted to try to land this for today's release, so I accepted the changes.

**Comment by Hipska** [comment]
@powersj my comments are still open and un-addressed..

**Comment by srebhan** [comment]
@Meceron can you please at least make those two strings const?

File: plugins/parsers/csv/parser_test.go [hunk 10]
```diff
@@ -221,6 +222,24 @@ func TestDelimiter(t *testing.T) {
...
+	p := &Parser{
+		HeaderRowCount: 0,
+		Delimiter:      "\u0000",
+		ColumnNames:    []string{"first", "second", "third"},
+		TimeFunc:       DefaultTime,
+	}
+	err := p.Init()
+	require.NoError(t, err)
+
+	testCSV := strings.Join([]string{"3.4", "70", "test_name"}, "\u0000")
```
[the remark this answers is not in the document]
Comment (powersj) [reply plugins/parsers/csv/parser_test.go:235]: I prefer the existing as it makes it very clear what is going on versus having to count and determine the end of a unicode value

**Comment by Hipska** [comment]
@srebhan This is what @powersj already did in a315d23.

**Comment by powersj** [comment]
> @Meceron can you please at least make those two strings const?

This was done already. I accepted your change that did this.

**Review by srebhan — APPROVED** [review]
Thanks for the nice improvement @Meceron! LGTM.

**Comment by powersj** [comment]
Merging as integration test is a known issue fixed on master

**Pull request merged**